semantic/utilities.go

```diff
@@ -0,0 +1,28 @@
+package semantic
+
+import (
+	"fmt"
+	"math/big"
+)
+
+func convertToBigIntOrPanic(str string) *big.Int {
+	if num, isNumber := convertToBigInt(str); isNumber {
+		return num
+	}
+
+	panic(fmt.Sprintf("failed to convert %s to a number", str))
```

Comment:
> An alternative could be to introduce a wrapping struct, so you can check if it is a number close to the comparison

could you expand or give an example of what you mean by that, just to make sure I understand the pattern you have in mind?

Reply:
> right now we behave like `Compare`

This panic can only be hit while parsing the version. So `(version) Compare(version) int` can still be error free because the parsing and error handling happened beforehand. This would only impact `(version) CompareStr(str) (int, error)` which seems reasonable to me because the `str` may not be a valid version.

This also seems more cohesive with the top level `Parse` which would look like:
```
func Parse(str string, ecosystem string) (Version, error) {
	//nolint:exhaustive // Using strings to specify ecosystem instead of lockfile types
	switch ecosystem {
	case "Alpine":
		return parseAlpineVersion(str)
	...
```

> all native version comparers handle "invalid" versions in a non-erroring way

What "native version comparers" are you referring to? In our use case specifically, we have no control over where the version strings comes from. And we'll be using this server side, so this will be a query of death for our service.

> this panic should never ever happen

Optimistic words :)

This assumes the regexes are always correct. If there's an update or a new ecosystem where a non-number can sneak through the regex would panic. A panic is dangerous because it is possible for deferred functions to deadlock or further corrupt internal or external state.  If submitted as is, we'd wrap `semantic.Parse(...)` in a recovery block to avoid this.

Reply:
> This panic can only be hit while parsing the version

nope, it's called on the compare path for Debian (`splitDebianDigitPrefix` -> `compareDebianVersions` -> `DebianVersion#Compare`)

> What "native version comparers" are you referring to?

I'm referring to the libraries and packages and code and everything that handles doing version comparisons in the native ecosystem for the particular version, which is what `semantic` is mirroring i.e. `org.apache.maven.artifact.versioning.ComparableVersion` in Maven, `Gem::Version` for Bundler, `packaging.version` in Python, `version_compare` in Composer, etc

The goal of `semantic` is to mirror both the spec and the implementation as closely as possible (in that order), so if a particular native version comparer rejected an input, I would have pursued `semantic` doing the same - at this point it has been so long since I implemented most of these that I don't want to swear by that and so yes its worth checking cause better safe than sorry, but I have high confidence and wanted to share some of the history behind the current structure.

> Optimistic words :)

I don't think they're _overly_ optimistic given the extensiveness of my test suite ;) (in saying that, I'd have a better argument if the test suite included a fuzzer 😅)

I don't disagree though that it's actually impossible, and yes I get your position more now knowing its in a server side context.

As some more background (and maybe as a small point of pride), I have done a _lot_ of testing as part of building all the comparators, involving a lot of manual and automatic crafting of version strings for these ecosystems, and tossing a lot of crap at both the native and `semantic` comparators to check their behaviour - despite that, yes a `panic` is death for a server, which trumps all the optimism in the world.

Reply:
>> An alternative could be to introduce a wrapping struct, so you can check if it is a number close to the comparison
>
> could you expand or give an example of what you mean by that, just to make sure I understand the pattern you have in mind?

Instead of `Parse(foo) (version, error)` you could do `Parse(foo) version`, where version has a functino `Valid() bool` or something like that. Then you could do the error handling in the place you actually use the version. Just a random thought, idk if this is the best idea here.